File: backend/src/controllers/orderController.js
const mongoose = require('mongoose');

const {
  CartItem,
  Order,
  OrderItem,
  Payment,
  Product,
} = require('../models');
const {
  getOrCreateActiveCart,
  buildCartSnapshot,
  clearCart,
  verifyStockLevels,
} = require('../services/cartService');
const paymentService = require('../services/paymentService');

exports.createOrder = async (req, res) => {
  const { shippingAddress, notes } = req.body;

  const cart = await getOrCreateActiveCart(req.user._id);
  const snapshot = await buildCartSnapshot(cart._id);

  if (!snapshot.items.length) {
    const error = new Error('Cart is empty');
    error.status = 400;
    throw error;
  }

  await verifyStockLevels(snapshot.items);

  const totalAmount = snapshot.totals.amount;
  const order = await Order.create({
    user: req.user._id,
    cart: cart._id,
    totalAmount,
    currency: 'INR',
    status: paymentService.isConfigured() ? 'pending_payment' : 'paid',
    shippingAddress,
    paymentIntentId: null,
    notes,
  });

  const orderItemsData = snapshot.items.map((item) => ({
    order: order._id,
    product: item.productId,
    productName: item.name,
    unitPrice: item.unitPrice,
    quantity: item.quantity,
    subtotal: item.subtotal,
  }));

  await OrderItem.insertMany(orderItemsData);

  for (const item of snapshot.items) {
    await Product.findByIdAndUpdate(item.productId, {
      $inc: { stock: -item.quantity },
    });
  }

  let razorpayOrder = null;

  if (paymentService.isConfigured()) {
    try {
      razorpayOrder = await paymentService.createRazorpayOrder({
        amount: totalAmount,
        currency: 'INR',
        receipt: order._id.toString(),
        notes: {
          userId: req.user._id.toString(),
        },
      });
      order.paymentIntentId = razorpayOrder.id;
      await order.save();
    } catch (err) {
      console.error('Failed to create Razorpay order', err);
      await Order.deleteOne({ _id: order._id });
      await OrderItem.deleteMany({ order: order._id });
      const error = new Error('Unable to initiate payment. Please try again later.');
      error.status = 502;
      throw error;
    }
  }

  const payment = await Payment.create({
    order: order._id,
    gateway: paymentService.isConfigured() ? 'razorpay' : 'manual',
    amount: totalAmount,
    currency: 'INR',
    status: paymentService.isConfigured() ? 'initiated' : 'captured',
    transactionId: paymentService.isConfigured() ? null : 'offline',
    rawResponse: razorpayOrder,
  });

  await clearCart(cart._id);

  const populated = await Order.findById(order._id)
    .populate('user', 'fullName email')
    .lean();

  res.status(201).json({
    success: true,
    data: {
      order: populated,
      items: orderItemsData,
      payment,
      razorpay: razorpayOrder
        ? {
            orderId: razorpayOrder.id,
            amount: razorpayOrder.amount,
            currency: razorpayOrder.currency,
            keyId: process.env.RAZORPAY_KEY_ID,
            name: 'Online Annavaram',
            description: `Temple pantry order ${populated._id}`,
          }
        : null,
    },
  });
};

exports.listOrders = async (req, res) => {
  const orders = await Order.find({ user: req.user._id })
    .sort({ createdAt: -1 })
    .lean();

  res.json({ success: true, data: orders });
};

exports.getOrder = async (req, res) => {
  const { id } = req.params;
  if (!mongoose.isValidObjectId(id)) {
    const error = new Error('Invalid order ID');
    error.status = 400;
    throw error;
  }

  const order = await Order.findOne({ _id: id, user: req.user._id })
    .populate('user', 'fullName email')
    .lean();

  if (!order) {
    const error = new Error('Order not found');
    error.status = 404;
    throw error;
  }

  const items = await OrderItem.find({ order: order._id }).lean();

  res.json({ success: true, data: { order, items } });
};

exports.listAllOrders = async (_req, res) => {
  const orders = await Order.find()
    .populate('user', 'fullName email role')
    .sort({ createdAt: -1 })
    .lean();

  res.json({ success: true, data: orders });
};

File: backend/src/controllers/orderController.py
from __future__ import annotations

import os
from typing import Any, Dict, List, Optional

from beanie.odm.fields import PydanticObjectId
from fastapi import HTTPException, status

from ..models import Order, OrderItem, Payment, Product, User
from ..services.cartService import (
    buildCartSnapshot,
    clearCart,
    getOrCreateActiveCart,
    verifyStockLevels,
)
from ..services.paymentService import (
    PaymentServiceError,
    createRazorpayOrder,
    isConfigured,
)


def _ensure_object_id(value: str) -> PydanticObjectId:
    try:
        return PydanticObjectId(value)
    except Exception:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid identifier")


def _serialize_order(order: Order, user: Optional[User] = None) -> Dict[str, Any]:
    payload = order.model_dump()
    payload["id"] = str(order.id)
    payload["user"] = {
        "id": str(user.id) if user else str(order.user),
        "fullName": user.fullName if user else None,
        "email": user.email if user else None,
    }
    return payload


def _serialize_payment(payment: Payment) -> Dict[str, Any]:
    payload = payment.model_dump()
    payload["id"] = str(payment.id)
    payload["order"] = str(payment.order)
    return payload


async def createOrder(*, user: User, shippingAddress: Dict[str, Any], notes: Optional[str]):
    cart = await getOrCreateActiveCart(user.id)
    snapshot = await buildCartSnapshot(cart.id)
    if not snapshot["items"]:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Cart is empty")

    try:
        await verifyStockLevels(snapshot["items"])
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc))

    total_amount = snapshot["totals"]["amount"]
    order_status = "pending_payment" if isConfigured() else "paid"

    order = Order(
        user=user.id,
        cart=cart.id,
        totalAmount=total_amount,
        currency="INR",
        status=order_status,
        shippingAddress=shippingAddress,
        paymentIntentId=None,
        notes=notes,
    )
    await order.insert()

    order_items_payload: List[Dict[str, Any]] = []
    for item in snapshot["items"]:
        order_items_payload.append(
            {
                "order": order.id,
                "product": PydanticObjectId(item["productId"]),
                "productName": item["name"],
                "unitPrice": item["unitPrice"],
                "quantity": item["quantity"],
                "subtotal": item["subtotal"],
            }
        )
    await OrderItem.insert_many([OrderItem(**payload) for payload in order_items_payload])

    for item in snapshot["items"]:
        await Product.find(Product.id == PydanticObjectId(item["productId"])).update(
            {"$inc": {"stock": -item["quantity"]}}
        )

    razorpay_order = None
    if isConfigured():
        try:
            razorpay_order = await createRazorpayOrder(
                amount=int(total_amount),
                currency="INR",
                receipt=str(order.id),
                notes={"userId": str(user.id)},
            )
            order.paymentIntentId = razorpay_order["id"]
            await order.save()
        except (PaymentServiceError, Exception) as exc:
            await Order.find(Order.id == order.id).delete()
            await OrderItem.find(OrderItem.order == order.id).delete()
            raise HTTPException(
                status_code=status.HTTP_502_BAD_GATEWAY,
                detail="Unable to initiate payment. Please try again later.",
            ) from exc

    payment = Payment(
        order=order.id,
        gateway="razorpay" if isConfigured() else "manual",
        amount=total_amount,
        currency="INR",
        status="initiated" if isConfigured() else "captured",
        transactionId=None if isConfigured() else "offline",
        rawResponse=razorpay_order,
    )
    await payment.insert()

    await clearCart(cart.id)

    order_payload = _serialize_order(order, user)
    items_payload = order_items_payload
    payment_payload = _serialize_payment(payment)

    return {
        "success": True,
        "data": {
            "order": order_payload,
            "items": items_payload,
            "payment": payment_payload,
            "razorpay": (
                {
                    "orderId": razorpay_order["id"],
                    "amount": razorpay_order["amount"],
                    "currency": razorpay_order["currency"],
                    "keyId": os.getenv("RAZORPAY_KEY_ID"),
                    "name": "Online Annavaram",
                    "description": f"Temple pantry order {order_payload['id']}",
                }
                if razorpay_order
                else None
            ),
        },
    }


async def listOrders(*, user: User):
    orders = await Order.find(Order.user == user.id).sort("-createdAt").to_list()
    return {"success": True, "data": [_serialize_order(order, user) for order in orders]}


async def getOrder(*, user: User, orderId: str):
    object_id = _ensure_object_id(orderId)
    order = await Order.find_one(Order.id == object_id, Order.user == user.id)
    if not order:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Order not found")
    items = await OrderItem.find(OrderItem.order == order.id).to_list()
    items_payload = [item.model_dump() | {"id": str(item.id), "order": str(item.order)} for item in items]
    return {"success": True, "data": {"order": _serialize_order(order, user), "items": items_payload}}


async def listAllOrders():
    orders = await Order.find_all().sort("-createdAt").to_list()
    return {"success": True, "data": [_serialize_order(order) for order in orders]}
